File: src/gui/itemnode/textfield/MoodItemNode.java
package gui.itemnode.textfield;

import audio.tagging.Metadata;
import gui.objects.picker.MoodPicker;
import gui.objects.popover.PopOver;
import gui.objects.popover.NodePos;
import util.access.V;
import util.parsing.Parser;
import static gui.objects.textfield.autocomplete.AutoCompletion.autoComplete;
import static main.App.APP;
import static util.functional.Util.filter;

/**
 * {@link gui.itemnode.textfield.TextFieldItemNode} for audio mood tagging values.<br/>
 * It provides two additional functionalities
 * <ul>
 * <li> Auto-completion from set of moods application is aware of
 * <li> Mood picker popup. The position of the picker popup can be customized.
 * </ul>
 */
public class MoodItemNode extends TextFieldItemNode<String> {

	/** The position for the picker to show on. */
	public final V<NodePos> pos = new V<>(NodePos.RightCenter);

	public MoodItemNode() {
		super(Parser.DEFAULT.toConverterOf(String.class));
		setEditable(true);
		autoComplete(this, p -> filter(APP.db.getStringPool().getStrings(Metadata.Field.MOOD.name()), t -> APP.db.getAutocompletionFilter().apply(t, p.getUserText())));
	}

	@Override
	void onDialogAction() {
		MoodPicker picker = new MoodPicker();
		PopOver<?> p = new PopOver<>(picker.getNode());
		p.detachable.set(false);
		p.setArrowSize(0);
		p.setArrowIndent(0);
		p.setCornerRadius(0);
		p.setAutoHide(true);
		p.setAutoFix(true);
		picker.onCancel = p::hide;
		picker.onSelect = mood -> {
			setValue(mood);
			p.hide();
		};
		picker.getNode().setPrefSize(800, 600);
		p.show(this, pos.get());
	}

}

File: src/player/main/sp/it/pl/ui/objects/textfield/MoodItemNode.kt
package sp.it.pl.ui.objects.textfield

import sp.it.pl.audio.tagging.Metadata.Field.MOOD
import sp.it.pl.main.APP
import sp.it.pl.ui.objects.autocomplete.AutoCompletion
import sp.it.pl.ui.objects.autocomplete.AutoCompletion.Companion.autoComplete

/** Text field for audio mood tagging values with a picker and auto-completion. */
class MoodItemNode: ValueTextField<String>() {

   init {
      isEditable = true
      autoComplete(this) { text ->
         APP.db.itemUniqueValuesByField[MOOD].orEmpty().filter { it.contains(text, true) }
      }
   }

   override fun onDialogAction() {
      AutoCompletion.of<String>(this)?.updateSuggestions("")
   }

}